File: apps/frontend/mfe-catalog/src/shared/api/types/lineage.ts
// Table Landing — GraphQL tableList types
export interface TableListItem {
  project: string;
  dataset: string;
  table: string;
  lastModified: string;                   // ISO datetime string (from publish_time)
  sizeBytes: number | null;               // from total_logical_size
  rowsWritten: number | null;             // from total_row_cnt (daily aggregate)
  writeMode: string | null;               // "append" | "fulldump" | "upsert" | null
}

export interface TableListResponse {
  items: TableListItem[];
  totalCount: number;
}

export interface TableRankingItem {
  tableId: string;       // "project.dataset.table_name"
  project: string;
  dataset: string;
  table: string;
  valueYesterday: number;
  value7dAvg: number;
  changePct: number;
  history7d: number[];   // 7 daily values, oldest → newest
}

// Basic Lineage Types (Legacy)
export interface LineageNode {
  id: string;
  type: string;
  name: string;
  job_id?: string | null;
  full_name?: string | null;
  owners?: string[];
  status?: string | null;
  enabled?: boolean | null;
}

export interface LineageEdge {
  source: string;
  target: string;
  io?: string | null;
}

export interface LineageGraphData {
  nodes: LineageNode[];
  edges: LineageEdge[];
}

// Health API Types
export interface FreshnessInfo {
  last_updated: string;
  sla: string;
  delay?: number | null;
}

export interface LastRunInfo {
  result: string;
  duration: string;
  ended_at: string;
}

export interface ExecutionInfo {
  mode: string;
  partition?: string | null;
}

export interface JobHealthData {
  freshness: FreshnessInfo;
  last_run: LastRunInfo;
  execution: ExecutionInfo;
}

export interface JobHealthResponse {
  job_id: string;
  updated_at: string;
  health: JobHealthData;
}

// Hybrid Lineage API Types
export interface InputTableInfo {
  id: string;
  name: string;
  storage_type?: string;
  read_mode?: string;
  freshness?: string | null;
  quality_status?: string | null;
  row_count?: number | null;
  owner?: string | null;
  criticality?: string | null;
}

export interface OutputTableInfo {
  id: string;
  name: string;
  storage_type?: string;
  write_mode?: string;
  recent_volume?: number | null;
  consumer_count: number;
  sla_status?: string | null;
}

export interface JobLineageHybridResponse {
  job_id: string;
  inputs: InputTableInfo[];
  outputs: OutputTableInfo[];
  graph: LineageGraphData;
}

// Impact Analysis API Types
export interface ImpactDownstreamEntry {
  depth: number;
  table: string;
  writer_jobs: string[];
  description?: string;
}

export interface ImpactAnalysisSummary {
  total_depth: number;
  total_downstream_tables: number;
  total_writer_jobs: number;
}

export interface ImpactAnalysisResponse {
  base_table: string;
  downstream: ImpactDownstreamEntry[];
  summary: ImpactAnalysisSummary;
}

File: apps/backend/lineage_manager_v2/app/api/v1/schemas/lineage.py
from typing import Any, Dict, List, Optional

from pydantic import BaseModel, ConfigDict


class LineageEdge(BaseModel):
    source_type: str  # 'job', 'data'
    source_name: str
    target_type: str
    target_name: str
    edge_type: str = "lineage"
    properties: Dict[str, Any] = {}


class LineageRegistration(BaseModel):
    edges: List[LineageEdge]


class NodeSchema(BaseModel):
    id: int
    type: str  # 'job', 'data'
    name: str  # fqn or job_id
    properties: Dict[str, Any] = {}


class EdgeSchema(BaseModel):
    id: int
    source: int
    target: int
    type: str
    properties: Dict[str, Any] = {}


class GraphResponse(BaseModel):
    nodes: List[NodeSchema]
    edges: List[EdgeSchema]


class InputTableInfo(BaseModel):
    id: str
    name: str
    storage_type: Optional[str] = None
    read_mode: Optional[str] = None
    freshness: Optional[str] = None
    quality_status: Optional[str] = None
    row_count: Optional[int] = None
    owner: Optional[str] = None
    criticality: Optional[str] = None


class OutputTableInfo(BaseModel):
    id: str
    name: str
    storage_type: Optional[str] = None
    write_mode: Optional[str] = None
    recent_volume: Optional[int] = None
    consumer_count: int = 0
    sla_status: Optional[str] = None


class LineageNode(BaseModel):
    id: str
    type: str
    name: str
    job_id: Optional[str] = None
    full_name: Optional[str] = None
    owners: List[str] = []
    status: Optional[str] = None
    enabled: Optional[bool] = None


class LineageGraphEdge(BaseModel):
    source: str
    target: str
    io: Optional[str] = None


class LineageGraphData(BaseModel):
    nodes: List[LineageNode]
    edges: List[LineageGraphEdge]


class JobLineageHybridResponse(BaseModel):
    job_id: str
    inputs: List[InputTableInfo]
    outputs: List[OutputTableInfo]
    graph: LineageGraphData


class MermaidNode(BaseModel):
    id: str
    type: str
    label: str
    properties: Dict[str, Any] = {}


class MermaidEdge(BaseModel):
    id: int
    source: str
    target: str
    type: str
    properties: Dict[str, Any] = {}


class MermaidMetadata(BaseModel):
    total_nodes: int
    depth: int
    truncated: bool


class MermaidGraphResponse(BaseModel):
    nodes: List[MermaidNode]
    edges: List[MermaidEdge]
    metadata: MermaidMetadata


# Impact Analysis Schemas

class ImpactDownstreamEntry(BaseModel):
    depth: int
    table: str
    writer_jobs: List[str] = []
    description: Optional[str] = None


class ImpactAnalysisSummary(BaseModel):
    total_depth: int
    total_downstream_tables: int
    total_writer_jobs: int


class ImpactAnalysisResponse(BaseModel):
    base_table: str
    downstream: List[ImpactDownstreamEntry]
    summary: ImpactAnalysisSummary
